### ISSUE TITLE
Directly destructuring awaited generic collapses to any

### ISSUE THREAD

**Issue by JasonKleban** [microsoft/TypeScript#29414]
New in 3.2.2(?) is this type mistake in `baz()` below.  `result` should be destructured to having type string, but instead it is `any` even though the unabbreviated syntaxes work correctly.

```ts
type responseType = { result: string }

declare function post<T = responseType>(): Promise<T>;

async function foo() {
    const response = await post(); // response : responseType  OK
    const result = response.result; // result : string OK
}

async function bar() {
    const response = await post(); // response : responseType OK
    const { result } = response; // result : string OK
}

async function baz() {
    const { result } = await post(); // result : any !!!!!!!!!!!
}
```

**Comment by RyanCavanaugh** [comment]
Simpler repro
```ts
declare function post<T = { result: string }>(): T;

function bar() {
  const response = post();
  const { result } = response;
}

function baz() {
  const { result } = post();
  result.subst_r; // should be error
}
```

**Comment by weswigham** [comment]
This is fixed in our latest nightlies, by #39081
